### PR TITLE
[Ready for Testing] DB-Code Refactoring & Moncole-Alternate support

### PR DESCRIPTION
**Pull request by 123FLO321**
# Description
Moved all queries to a new class.
This allows easy support for different Scanners and even DB types.
Currently RocketMap [MySQL] and the Monocle-Alternate [MySQL and PgSQL] fork are implemented.
The DB Type can be set via in `config.php`. `["rocketmap", "monocle-alt-mysql" or "monocle-alt-pgsql"]`.

Run CreateStatsRocketMap.sql or CreateStatsMonocle.sql for counts to work.

Requires PR 27 Moncole-Alternate
Recommended Settings for Moncole-Alternate:
```
KEEP_GYM_HISTORY = True
CLEANUP_SIGHTINGS_OLDER_THAN_X_HR = 168.0 (or higher | min: 24.0)
CLEANUP_FORT_SIGHTINGS_OLDER_THAN_X_HR = -1.0 (or 168.0 or higher)
CLEANUP_GYM_DEFENDERS_OLDER_THAN_X_HR = -1.0 (or 168.0 or higher)
RAIDERS_PER_GYM = 0.1 (target update interval of ~60s)
```

## ToDo
- [ ] Test RocketMap
- [ ] Test Moncole-Alternate MySQL
- [ ] Test Moncole-Alternate PgSQL
- [x] Implement the Trainer-Page for  Moncole-Alternate once it stores levels in the dB
- [x] Implement Gymhistory for  Moncole-Alternate

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by javisabalete** [comment]
~~Tested with RocketMap and works as expected~~.

Requested small changes

File: core/process/queries/QueryManagerRocketmap.php [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+	
+	function getTotalGymsForTeam($team_id) {
+		$req = "SELECT COUNT(DISTINCT(gym_id)) AS total FROM gym WHERE team_id = '".$team_id."'";
+		$result = $this->mysqli->query($req);
+		$data = $result->fetch_object();
+		return $data;
+	}
+	
+	function getRecentAll() {
+		$req = "SELECT DISTINCT pokemon_id, encounter_id, disappear_time, last_modified, (CONVERT_TZ(disappear_time, '+00:00', '".$this->time_offset."')) AS disappear_time_real,
```
Comment (javisabalete) [inline core/process/queries/QueryManagerRocketmap.php]: `$this->time_offset` must be `self::time_offset`

File: core/process/queries/QueryManagerRocketmap.php [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+		if ($result->num_rows > 0) {
+			while ($row = $result->fetch_object()) {
+				$data[] = $row;
+			}
+		}
+		return $data;
+	}
+	
+	function getRecentMythic($mythic_pokemons) {
+		$req = "SELECT DISTINCT pokemon_id, encounter_id, disappear_time, last_modified, (CONVERT_TZ(disappear_time, '+00:00', '".$this->time_offset."')) AS disappear_time_real,
```
Comment (javisabalete) [inline core/process/queries/QueryManagerRocketmap.php]: Here too, `$this->time_offset` must be `self::time_offset`

**Review by javisabalete — CHANGES_REQUESTED** [review]

File: core/process/queries/QueryManagerRocketmap.php (hunk 2 above)
Comment (123FLO321) [reply core/process/queries/QueryManagerRocketmap.php]: fixed

**Review by javisabalete — APPROVED** [review]

File: core/process/queries/QueryManager.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	// Raids
+	abstract public function getAllRaids($page);
+
+	// Trainers
+	abstract public function getTrainers($trainer_name, $team, $page, $ranking);
+	abstract public function getTrainerLevelCount($team_id);
+
+	// Cron
+	abstract public function getPokemonCountsActive();
+	abstract public function getPoekmonCountsLastDay();
```
Comment (michikrug) [inline core/process/queries/QueryManager.php]: Typo

File: core/process/queries/QueryManager.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	// Single Pokemon
+	abstract public function getGymsProtectedByPokemon($pokemon_id);
+	abstract public function getPokemonLastSeen($pokemon_id);
+	abstract public function getTop50Pokemon($pokemon_id, $top_order_by, $top_direction);
+	abstract public function getTop50Trainers($pokemon_id, $best_order_by, $best_direction);
+	abstract public function getPokemonHeatmap($pokemon_id, $start, $end);
+	abstract public function getPokemonGraph($pokemon_id);
+	abstract public function getPokemonLive($pokemon_id, $ivMin, $ivMax, $inmap_pokemons);
+	abstract public function getPokemonSliederMinMax();
```
Comment (michikrug) [inline core/process/queries/QueryManager.php]: Typo

File: core/process/queries/QueryManager.php (hunk 3 above)
Comment (michikrug) [inline core/process/queries/QueryManager.php]: Typo

File: core/cron/pokedex_rarity.cron.php [hunk 5]
```diff
@@ -2,16 +2,12 @@
 
 // get pokemon from last day to calculate the rarity for last 24h
 // use last disappear_time as a starting point to overcome scan downtimes
-$req = "SELECT pokemon_id, COUNT(*) AS spawns_last_day
-		FROM pokemon
-		WHERE disappear_time >= (SELECT MAX(disappear_time) FROM pokemon) - INTERVAL 1 DAY
-		GROUP BY pokemon_id
-		ORDER BY pokemon_id ASC";
-$result = $mysqli->query($req);
+$data = $manager->getPoekmonCountsLastDay();
```
Comment (michikrug) [inline core/cron/pokedex_rarity.cron.php]: Typo

File: core/process/queries/QueryManagerPostgresqlMonocleAlternate.php [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+					LIMIT 5000";
+		$result = pg_query($this->db, $req);
+		$spawns = array();
+		while ($data = pg_fetch_object($result)) {
+			$spawns[] = $data;
+		}
+		return $spawns;
+	}
+
+	public function getPokemonSliederMinMax() {
```
Comment (michikrug) [inline core/process/queries/QueryManagerPostgresqlMonocleAlternate.php]: Typo

File: core/process/queries/QueryManagerMysqlMonocleAlternate.php [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+					LIMIT 5000";
+		$result = $this->mysqli->query($req);
+		$spawns = array();
+		while ($data = $result->fetch_object()) {
+			$spawns[] = $data;
+		}
+		return $spawns;
+	}
+
+	public function getPokemonSliederMinMax() {
```
Comment (michikrug) [inline core/process/queries/QueryManagerMysqlMonocleAlternate.php]: Typo

File: core/process/queries/QueryManagerMysqlMonocleAlternate.php [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+		$req = "SELECT pokemon_id, COUNT(*) as total FROM sightings WHERE expire_timestamp >= UNIX_TIMESTAMP() GROUP BY pokemon_id";
+		$result = $this->mysqli->query($req);
+		$counts = array();
+		while ($data = $result->fetch_object()) {
+			$counts[$data->pokemon_id] = $data->total;
+		}
+		return $counts;
+	}
+
+	public function getPoekmonCountsLastDay() {
```
Comment (michikrug) [inline core/process/queries/QueryManagerMysqlMonocleAlternate.php]: Typo

File: core/process/queries/QueryManagerPostgresqlMonocleAlternate.php [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+		$req = "SELECT pokemon_id, COUNT(*) as total FROM sightings WHERE expire_timestamp >= EXTRACT(EPOCH FROM NOW()) GROUP BY pokemon_id";
+		$result = pg_query($this->db, $req);
+		$counts = array();
+		while ($data = pg_fetch_object($result)) {
+			$counts[$data->pokemon_id] = $data->total;
+		}
+		return $counts;
+	}
+
+	public function getPoekmonCountsLastDay() {
```
Comment (michikrug) [inline core/process/queries/QueryManagerPostgresqlMonocleAlternate.php]: Typo

**Review by michikrug — CHANGES_REQUESTED** [review]
Typos

**Comment by 123FLO321** [comment]
fixed @michikrug

File: core/cron/nests.cron.php [hunk 10]
```diff
@@ -17,8 +17,17 @@
 	$nests['lng'] = $data->longitude;
 	$starttime = $data->latest_seen - $data->duration;
 
-	$nests['st'] = date("i",$starttime);;
-	$nests['et'] = date("i",$data->latest_seen);;
+	$startdate = date("i",$starttime);
+	$enddate = date("i",$data->latest_seen);
+
+	if ($enddate < $startdate) { // Smaller number first
```
Comment (michikrug) [inline core/cron/nests.cron.php]: I don't think this is correct, since we can have spawns from, e.g. :50 to :20

**Review by michikrug — CHANGES_REQUESTED** [review]

File: core/cron/nests.cron.php (hunk 10 above)
Comment (123FLO321) [reply core/cron/nests.cron.php]: oh yeah right.
Woops

**Review by Obihoernchen — CHANGES_REQUESTED** [review]
I get the following error:
```
PHP Fatal error:  Declaration of QueryManagerPostgresqlMonocleAlternate::getNestData($time, $minLatitude, $maxLatitude, $minLongitude, $maxLongitude) must be compatible with QueryManager::getNestData() in .../core/process/queries/QueryManagerPostgresqlMonocleAlternate.php on line 708
```

**Review by Obihoernchen — APPROVED** [review]
Seems to work fine with RM

**Comment by Obihoernchen** [comment]
Now in #371

**Pull request closed without merging**